src/Chainweb/Rosetta/Util.hs

```diff
@@ -0,0 +1,686 @@
+      tx = f cr ops
+  pure $ T2 nextIdx (acc txs tx)
+  where
+    -- Returns logs at given index and the next index to try
+    getLogs
+        :: Int
+        -> OperationType
+        -> Either String (T2 Int [UnindexedOperation])
+    getLogs i otype = do
+      (tid,l) <- peekLog i logs
+      let opsF = map (operation Successful otype tid) l
+      pure $ T2 (succ i) opsF
+
+    getTransferLogs
+        :: Int
+        -> Maybe TxId
+        -> Either String (T2 Int [UnindexedOperation])
+    getTransferLogs expected actual =
+      case actual of
+        Nothing -> noTransferLogs
+        Just actualTid -> peekNextTxId >>= (isCoinTx actualTid)
+      where
+        noTransferLogs = pure $ T2 expected []
+        transferLogs = getLogs expected TransferOrCreateAcct
+        peekNextTxId = fmap fst (peekLog expected logs)
+        isCoinTx aTid eTid
+          | (eTid == aTid) = transferLogs
+          | otherwise = noTransferLogs
+
+
+-- TODO: Max limit of tx to return at once.
+--       When to do pagination using /block/transaction?
+-- |
+-- | Matches all transactions in a non-genesis block to their coin contract logs.
+--   The first transaction in non-genesis blocks is the coinbase transaction.
+--   Each transactions that follows has (1) logs that fund the transaction,
+--   (2) optional tx specific coin contract logs, and (3) logs that pay gas.
+-- |
+nonGenesisTransactions
+    :: Map TxId [AccountLog]
+    -> CoinbaseCommandResult
+    -> V.Vector (CommandResult Hash)
+    -> NonGenesisMatchFunction [Transaction]
+nonGenesisTransactions logs initial rest =
+  NonGenesisMatchFunction $
+    overwriteError RosettaMismatchTxLogs $
+      nonGenesisTransactions'
+      logs _crTxId rosettaTransaction initial rest
+
+nonGenesisTransactions'
+    :: Map TxId [AccountLog]
+    -> (a -> Maybe TxId)
+    -> (a -> [Operation] -> b)
+    -> a
+    -> V.Vector a
+    -> Either String [b]
+nonGenesisTransactions' logs getTxId f initial rest = do
+  let initIdx = 0
+  initLog <- peekLog initIdx logsVector
+  initTx <- nonGenesisCoinbase initLog getTxId f initial
+  T2 _ ts <- foldM match (defAcc initIdx initTx) rest
+  pure $ DList.toList ts
+  where
+    logsVector = V.fromList $ M.toAscList logs   -- O(1) lookup by index
+    match = gasTransactionAcc logsVector getTxId f DList.snoc
+    defAcc i tx = T2 (succ i) (DList.singleton tx)
+
+
+-- |
+-- | Matches a single non-genesis transaction to its coin contract logs
+-- |
+nonGenesisTransaction
+    :: Map TxId [AccountLog]
+    -> CoinbaseCommandResult
+    -> V.Vector (CommandResult Hash)
+    -> RequestKey
+    -> NonGenesisMatchFunction Transaction
+nonGenesisTransaction logs initial rest target =
+  NonGenesisMatchFunction $
+    (noteOptional RosettaTxIdNotFound .
+     overwriteError RosettaMismatchTxLogs) $
+       nonGenesisTransaction'
+       logs _crReqKey _crTxId rosettaTransaction initial rest target
+
+nonGenesisTransaction'
```

Comment:
Also same comment as above regarding polymorphism and roles of `a` and `b`, it makes it hard to understand what is being accomplished here.

Reply:
Got rid of the `getXXX` indirection. And yes part of the indirection was there for the tests because debugging is easier with `Either String rosettaTx`. But I added the conversion of the `Either String rosettaTx` to the `matchLogs` function.

@slpopejoy @emilypi I tried to simplify the matching logic and cut down on the polymorphism. See this commit for exactly what changed: https://github.com/kadena-io/chainweb-node/pull/1050/commits/ee355f55305ed9a6637e9603572189dab48b1172